package.json

```diff
@@ -19,7 +19,9 @@
     "clean": "rimraf dist",
     "prepublish": "safe-publish-latest && npm run build"
   },
-  "dependencies": {},
+  "dependencies": {
+    "express": "^4.16.3"
```

Comment:
express would need to be a peer dep - this addition is a breaking change, by forcing compatibility with a specific version range of express.

Reply:
This is a tricky situation.  This version of express should have no bearing on the version of express that the app this lib is installed in is using to serve the application.  On the other hand, the version of express is important to anything that seeks to add behavior to the control plane (whether that is the application or some other utility or plugin).

Another alternative would be to make the control plane less general and tie into the existing pool-hall event emitters.  I could imagine having like

```
POST /sendWorker/:id

{"act": "heapdump"}
```

which would result in (supervisor)

```
poolHall.workers[id].send({ poolHallControlMsg: true, msg: { act: 'heapdump'} })
```

and

```
poolHall.worker.on('message', (msg) => {
  if (poolHallControlMsg) {
    if (msg.msg === 'heapdump') {
      ...
```

This API could be implemented using an internal express version or just raw node, making the versioning not be an issue, maybe?

Reply:
if poolHall isn't currently dependent on express, but "startControlPane" is, then perhaps it should be a separate package?

Reply:
This could work, but it pushes the express dependency issue to that package instead.  I think offering a standard control plane is generally desirable, as it would allow other plugins build around it.

What we really want is a way to send messages to the supervisor that result in actions happening either on the supervisor or on the workers.  Express is mostly irrelevant here, except that writing it without express would be somewhat more painful (and we'd probably have to write our own API that is very similar for plugins to use).

Reply:
Ok, concretely, what I could see is that we're at pool-hall 0.0.3, we've added a dep on express "^4.16.3" and we expose the instance of express as `poolHall.controlPlane` or whatever.

We have an application called `hypertension` that serves http requests.  For any number of reasons, it's on express "3.21.2" for the http requests it serves to users via an instance called `server`.  It has a dependency on pool-hall "^0.0.3".  This means that `poolHall.controlPlane` is `express=^4.16.3` and that API should be used for implementing actions on the control plane.  `server` is `express=3.21.2` and it continues doing its own thing.  It's kind of annoying, because we have two separate copies of different versions of `express` hiding out in different parts of the `node_modules` tree, but npm can handle this.

If `hypertension` uses a semver compatible version of express for itself, then the dependency can be shared, but it's still the `pool-hall` version that's defining the API version of express for `controlPlane`.

If we want to upgrade `pool-hall` to express 5.x.x, then this would be a breaking pool-hall version change, as our effective API has changed.  If we upgraded the express version `hypertension`, this would have no bearing on `pool-hall` or any actions built on the control plane.

Is this accurate or am I missing something?

Reply:
Yes, that sounds accurate.